File: core/src/com/infeez/simple/ResourceSingleton.java
package com.infeez.simple;


import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.TextureRegion;

import java.util.*;

public class ResourceSingleton {

    private static Texture cellsTexture;
    private static TextureRegion blackCell;
    private static TextureRegion whiteCell;

    private static Texture checkersTexture;
    private static TextureRegion blackChecker;
    private static TextureRegion whiteChecker;

    private static final Set<Integer> ids = new HashSet<>(1);

    public static Texture getCellsTexture(){
        if(cellsTexture == null){
            cellsTexture = new Texture(Gdx.files.internal("cells.jpg"));
        }
        return cellsTexture;
    }

    public static Texture getCheckersTexture(){
        if(checkersTexture == null){
            checkersTexture = new Texture(Gdx.files.internal("checkers.png"));
        }
        return checkersTexture;
    }

    public static TextureRegion getBlackCell(){
        if(blackCell == null){
            blackCell = new TextureRegion(getCellsTexture(), 50, 0, 50, 50);
        }
        return blackCell;
    }

    public static TextureRegion getWhiteCell(){
        if(whiteCell == null){
            whiteCell = new TextureRegion(getCellsTexture(), 0, 0, 50, 50);
        }
        return whiteCell;
    }

    public static TextureRegion getBlackChecker(){
        if(blackChecker == null){
            blackChecker = new TextureRegion(getCheckersTexture(), 50, 0, 50, 50);
        }
        return blackChecker;
    }

    public static TextureRegion getWhiteChecker(){
        if(whiteChecker == null){
            whiteChecker = new TextureRegion(getCheckersTexture(), 0, 0, 50, 50);
        }
        return whiteChecker;
    }

    public static int getUniqueId(){
        int rnd = new Random().nextInt(32 * 32);
        if(!ids.add(rnd)) {
            return getUniqueId();
        }
        return rnd;
    }

    public static void dispose(){
        if(cellsTexture == null) {
            return;
        }
        cellsTexture.dispose();
    }
}

File: core/src/main/kotlin/com/infeez/simple/ResourceSingleton.kt
package com.infeez.simple

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.infeez.simple.utils.BoardConfig
import java.util.concurrent.atomic.AtomicInteger

object ResourceSingleton {
    private var cellsTexture: Texture? = null
    private var blackCell: TextureRegion? = null
    private var whiteCell: TextureRegion? = null
    private var checkersTexture: Texture? = null
    private var blackChecker: TextureRegion? = null
    private var whiteChecker: TextureRegion? = null

    private val nextId = AtomicInteger()

    @JvmStatic
    fun getCellsTexture(): Texture {
        return cellsTexture ?: Texture(Gdx.files.internal("cells.jpg")).also {
            cellsTexture = it
        }
    }

    @JvmStatic
    fun getCheckersTexture(): Texture {
        return checkersTexture ?: Texture(Gdx.files.internal("checkers.png")).also {
            checkersTexture = it
        }
    }

    @JvmStatic
    fun getBlackCell(): TextureRegion {
        return blackCell ?: TextureRegion(
            getCellsTexture(),
            0,
            0,
            BoardConfig.TEXTURE_REGION_SIZE,
            BoardConfig.TEXTURE_REGION_SIZE,
        ).also {
            blackCell = it
        }
    }

    @JvmStatic
    fun getWhiteCell(): TextureRegion {
        return whiteCell ?: TextureRegion(
            getCellsTexture(),
            BoardConfig.TEXTURE_REGION_SIZE,
            0,
            BoardConfig.TEXTURE_REGION_SIZE,
            BoardConfig.TEXTURE_REGION_SIZE,
        ).also {
            whiteCell = it
        }
    }

    @JvmStatic
    fun getBlackChecker(): TextureRegion {
        return blackChecker ?: TextureRegion(
            getCheckersTexture(),
            BoardConfig.TEXTURE_REGION_SIZE,
            0,
            BoardConfig.TEXTURE_REGION_SIZE,
            BoardConfig.TEXTURE_REGION_SIZE,
        ).also {
            blackChecker = it
        }
    }

    @JvmStatic
    fun getWhiteChecker(): TextureRegion {
        return whiteChecker ?: TextureRegion(
            getCheckersTexture(),
            0,
            0,
            BoardConfig.TEXTURE_REGION_SIZE,
            BoardConfig.TEXTURE_REGION_SIZE,
        ).also {
            whiteChecker = it
        }
    }

    @JvmStatic
    fun getUniqueId(): Int {
        return nextId.getAndIncrement()
    }

    @JvmStatic
    fun dispose() {
        cellsTexture?.dispose()
        checkersTexture?.dispose()

        cellsTexture = null
        checkersTexture = null
        blackCell = null
        whiteCell = null
        blackChecker = null
        whiteChecker = null
    }
}
